src/tests/Microsoft.PowerFx.Core.Tests/UserDefinedTypeTests.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.PowerFx.Core.Public.Types;
+using Microsoft.PowerFx.Core.Tests.Helpers;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Syntax;
+using Xunit;
+
+namespace Microsoft.PowerFx.Core.Tests
+{
+    public class UserDefinedTypeTests : PowerFxTest
+    {
+        private static readonly ReadOnlySymbolTable _primitiveTypes = ReadOnlySymbolTable.PrimitiveTypesTableInstance;
+
+        [Theory]
+
+        // check record, table types with primitive types
+        [InlineData("Point = Type({ x: Number, y: Number })", "![x:n,y:n]", true)]
+        [InlineData("Points = Type([{ x: Number, y: Number }])", "*[x:n,y:n]", true)]
+        [InlineData("Person = Type({ name: Text, dob: Date })", "![name:s,dob:D]", true)]
+        [InlineData("People = Type([{ name: Text, isReady: Boolean }])", "*[name:s,isReady:b]", true)]
+        [InlineData("Heights = Type([Number])", "*[Value:n]", true)]
+        [InlineData("Palette = Type([Color])", "*[Value:c]", true)]
+        
+        // type alias
+        [InlineData("DTNZ = Type(DateTimeTZInd)", "Z", true)]
+
+        // nested record types
+        [InlineData("Nested = Type({a: {b: DateTime, c: {d: GUID, e: Hyperlink}}, x: Time})", "![a:![b:d, c:![d:g, e:h]], x:T]", true)]
+
+        // invalid types
+        [InlineData("Pics = Type([Image])", "*[Value:i]", false)]
+        public void TestUserDefinedType(string typeDefinition, string expectedDefinedTypeString, bool isValid)
+        {
+            var parseOptions = new ParserOptions
+            {
+                AllowParseAsTypeLiteral = true
+            };
+
+            UserDefinitions.ProcessUserDefinitions(typeDefinition, parseOptions, out UserDefinitionResult results, nameResolver: _primitiveTypes);
+
+            if (isValid) 
+            {
+                Assert.NotEmpty(results.DefinedTypes);
+                var resultType = results.DefinedTypes.First().Type._type;
+                Assert.Equal(TestUtils.DT(expectedDefinedTypeString), resultType);
+            }
+            else
+            {
+                Assert.Empty(results.DefinedTypes);
+                Assert.NotEmpty(results.Errors);
+            }
+        }
+
+        [Theory]
```

Comment:
>Theory [](http://example.com/codeflow?start=9&length=6)

// redeclare 
X= Type({ f:Number f:Number});

Reply:
great catch will add that